### PR TITLE
[macOS] Migrate @NSApplicationMain attribute to @main

### LINKED ISSUE #143044: AppDelegate.swift:4:1 'UIApplicationMain' is deprecated; this is an error in Swift 6
In Xcode 15.3 beta I see a deprecation warning in a newly created Flutter iOS Xcode project.  `@UIApplicationMain` was renamed to `@main` in Swift 5.3. 
```
AppDelegate.swift:4:1 'UIApplicationMain' is deprecated; this is an error in Swift 6
```
https://github.com/flutter/flutter/blob/db141ecd28e166b2ba0f4903c3fd9fb13e4eb1f3/packages/flutter_tools/templates/app_shared/ios-swift.tmpl/Runner/AppDelegate.swift#L4

We already require Xcode 14, and Swift 5.3 was introduced in Xcode 12.

I bet the same is needed for `@NSApplicationMain` but I didn't check.
https://github.com/flutter/flutter/blob/db141ecd28e166b2ba0f4903c3fd9fb13e4eb1f3/packages/flutter_tools/templates/app_shared/macos.tmpl/Runner/AppDelegate.swift#L4

I'm not sure if there are other implications to using `@main` instead of `@UIApplicationMain` (scene delegates?)

### PR DESCRIPTION
**Pull request by loic-sharma**
This migrates Flutter to use the `@main` attribute introduced in Swift 5.3. The `@NSApplicationMain` attribute is deprecated and will be removed in Swift 6. See: https://github.com/apple/swift-evolution/blob/main/proposals/0383-deprecate-uiapplicationmain-and-nsapplicationmain.md

This change is split into two commits:

1. https://github.com/flutter/flutter/commit/a508d3e5038c36849448f6b0b2567e8d416842fa - This updates the macOS app template and adds a migration to replace `@NSApplicationMain` uses with `@main`. 
2. https://github.com/flutter/flutter/commit/f43482786edafc48ee90bea607704c4a0c8cb96f - I ran `flutter run -d macos` on each Flutter macOS app in this repository to verify the app migrates and launches successfully.

Follow-up to https://github.com/flutter/flutter/pull/146707
Fixes https://github.com/flutter/flutter/issues/143044

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

File: packages/flutter_tools/lib/src/macos/migrations/nsapplicationmain_deprecation_migration.dart [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+const String _appDelegateFileAfter = r'''
+@main
+class AppDelegate''';
+
+/// Replace the deprecated `@NSApplicationMain` attribute with `@main`.
+///
+/// See:
+/// https://github.com/apple/swift-evolution/blob/main/proposals/0383-deprecate-uiapplicationmain-and-nsapplicationmain.md
+class NSApplicationMainDeprecationMigration extends ProjectMigrator {
```
Comment (hellohuanlin) [inline packages/flutter_tools/lib/src/macos/migrations/nsapplicationmain_deprecation_migration.dart:21]: im wondering if it's possible to reuse the ios migration?

**Review by hellohuanlin — APPROVED** [review]
Thanks for doing this! 

@jmagman One thing in my mind is whether this kind of migration scripts will get out of hand in the future? Because keeping track of the delta can be tedious. 

Strawman idea - what if we just regenerate the whole thing? This way we can have a simple one-for-all migration script that simply detects some sort of version change.

**Comment by jmagman** [comment]
> Strawman idea - what if we just regenerate the whole thing? This way we can have a simple one-for-all migration script that simply detects some sort of version change.

We can't regenerate files Flutter doesn't "own", at least not easily, or we would blow away the user changes.  For example in https://github.com/flutter/flutter/pull/146707/files#r1566254461, if we had regenerated the AppDelegate the user would have lost all their changes.

https://github.com/flutter/flutter/blob/224099872edbb1908466ee4de47c3aa56a58258e/examples/platform_channel_swift/ios/Runner/AppDelegate.swift#L22-L23

The Xcode project files are another example: users are _supposed_ to edit their Xcode project in some cases (as you know from the validator), so we can't just regenerate it for them without keeping track of the deltas in a much more complicated, error-prone way.

There was an attempt to do something much more complicated with https://github.com/flutter/flutter/wiki/Flutter-migrate, (see https://flutter.dev/go/flutter-migrate and https://github.com/flutter/flutter/issues/40460), but it was extremely complicated and required users running a tool explicitly, and allowing rollbacks.  There wasn't much appetite to finish up that project since these incremental migrations that essentially just string match (the first one was introduced in https://github.com/flutter/flutter/pull/51453) were doing a decent enough job.  Before that, users would just get build errors and have to figure out how to fix them themselves.

File: packages/flutter_tools/lib/src/macos/migrations/nsapplicationmain_deprecation_migration.dart (hunk 1 above)
Comment (loic-sharma) [reply packages/flutter_tools/lib/src/macos/migrations/nsapplicationmain_deprecation_migration.dart:21]: They're very similar but subtly different - the affected file is different, the log messages are slightly different (due to differences in path), the attribute is different. The migration logic would need to be more a little more complicated to support both platforms. Personally I think this duplication results in simple code that's easier to maintain.

File: dev/integration_tests/flavors/macos/Runner/AppDelegate.swift [hunk 2]
```diff
@@ -5,7 +5,7 @@
 import Cocoa
 import FlutterMacOS
 
-@NSApplicationMain
+@main
```
Comment (loic-sharma) [inline dev/integration_tests/flavors/macos/Runner/AppDelegate.swift:8]: @cbracken, this app appears to be busted before this migration.

**Review by jmagman — APPROVED** [review]
LGTM!

File: dev/integration_tests/flavors/macos/Runner/AppDelegate.swift (hunk 2 above)
Comment (cbracken) [reply dev/integration_tests/flavors/macos/Runner/AppDelegate.swift:8]: Oh interesting -- did someone disable the flavours test I wonder?

Comment (jmagman) [reply dev/integration_tests/flavors/macos/Runner/AppDelegate.swift:8]: It never passed: https://github.com/flutter/flutter/issues/119782
https://github.com/flutter/flutter/pull/125581

Comment (cbracken) [reply dev/integration_tests/flavors/macos/Runner/AppDelegate.swift:8]: That would do it then :)

**Pull request merged**